### DISCUSSION TITLE
Are you using CMake projects from the build directory instead of installing them?

### DISCUSSION THREAD

**Discussion by traversaro** [robotology/community#695] [Polls]
Back in time (like <2015), it was common practice in the iCub-related groups to find CMake projects (for example `YCM`, `YARP`, `ICUB`) by setting the `<PackageName>_DIR` environment variable to point to the build directory of the project. In this way, any call to `find_package(<PackageName>)` would find the libraries contained in the build directory, as opposed to the one installed, permitting users to avoid to install a library when they modified it. 

For example, this workflow is the one still documented in the ["Installation on icub-head installing manually each package"](https://icub-tech-iit.github.io/documentation/sw_installation/icub_head_manual/#compile-icub).

However, time has passed and the use of the robotology-superbuild reduced the use of this pattern. Maintaining this workflow in CMake files such as YCM's [`InstallBasicPackageFiles`](https://github.com/robotology/ycm/blob/v0.16.2/modules/InstallBasicPackageFiles.cmake) or other CMake files is hard to maintain and test (due to lack of testing this workflow is broken for many projects).

So, personally I would like to stop caring for this workflow. However, I wanted to understand if someone still uses this workflow. Please vote in this poll, and if you voted `Yes` or `I am not sure`, feel free to contact me either commenting here or with other means, thanks!

**Comment by traversaro** [comment]
fyi @robotology/everyone

**Reply by traversaro** [reply]
@robotology/everyone it seems that 4 people voted yes, it would be great if the people that voted yes could explain the workflow you are using in the comments, as @pattacini did, or eventually reaching me personally, thanks!

**Comment by reafrancesco** [comment]
Can we discuss this in details so that I can explain our position?
Francesco Rea


From: Silvio Traversaro ***@***.***>
Sent: martedì 9 aprile 2024 13:46
To: robotology/community ***@***.***>
Cc: Francesco Rea ***@***.***>; Team mention ***@***.***>
Subject: [Robotology] Are you using CMake projects from the build directory instead of installing them? (Discussion #695)


Back in time (like <2015), it was common practice to find CMake projects (for example YCM, YARP, ICUB) by setting the <PackageName>_DIR environment variable to point to the build directory of the project. In this way, any call to find_package(<PackageName>) would find the libraries contained in the build directory, as opposed to the one installed, permitting users to avoid to install a library when they modified it.

For example, this workflow is the one still documented in the "Installation on icub-head installing manually each package"<https://icub-tech-iit.github.io/documentation/sw_installation/icub_head_manual/#compile-icub>.

However, time has passed and the use of the robotology-superbuild reduced the use of this pattern. Maintaining this workflow in CMake files such as YCM's InstallBasicPackageFiles<https://github.com/robotology/ycm/blob/v0.16.2/modules/InstallBasicPackageFiles.cmake> or other CMake files is hard to maintain and test (due to lack of testing this workflow is broken for many projects).

So, personally I would like to stop caring for this workflow. However, I wanted to understand if someone still uses this workflow. Please vote in this poll, and if you voted Yes or I am not sure, feel free to contact me either commenting here or with other means, thanks!

**Comment by pattacini** [comment]
I'm mostly using `robotology-superbuild` with the install pattern. However, on Windows, for quick tests, I also rely on `<PackageName>_DIR` just for YCM, YARP, ICUB, and ICUBcontrib.

**Comment by PeterBowman** [comment]
What is actually going to change in InstallBasicPackageFiles.cmake if support for this workflow is dropped?

**Reply by traversaro** [reply]
If no one was using that, I would simply stopped producing a `<Project>Config.cmake` file in the build directory, i.e. removing https://github.com/robotology/ycm/blob/adccba6728c101a3d1bc7e532e26a1902bb22de9/modules/InstallBasicPackageFiles.cmake#L624C11-L637 . However, that code is only part of what is necessary for a project to support the workflow of finding the package from the build directory, as you also need to make sure that the structure in your build directory mirrors the one in your install directory.

As it seems that the workflow is indeed used, what I would probably eventually do is to add an option (so more complexity, not less :D ) in InstallBasicPackageFiles to optionally avoid to produce a `<Project>Config.cmake` file in the build directory.

**Comment by arrenglover** [comment]
@traversaro on my machine (ubuntu22) I'm creating a local folder to install to `~/install` and setting 

`export CMAKE_PREFIX_PATH=$CMAKE_PREFIX_PATH:~/install`

`export PATH=$PATH:~/install/bin`

then I set the install location of `YCM`, `YARP`, `event-driven` to `~/install` and then `make install`

I'm guessing this method would use `InstallBasicPackageFiles` as I know I have it in `event-driven` CMakeLists

**Reply by traversaro** [reply]
Thanks @arrenglover, yes that is the method **not** using the build directory (and what I was assuming most users are using).

**Comment by traversaro** [comment]
Ok, a bit or recap: given that there are some users of this feature, as mentioned in https://github.com/orgs/robotology/discussions/695#discussioncomment-9068984 think the best course of action is to modify `InstallBasicPackageFiles` to permit projects that do not support using the CMake package from build directory to explicitly do so (so that it is not possible that users accidentally find the `*Config.cmake` file in the build directory, and try to use it failing), while leave projects in which users are used to consume the library from the `build` (`YCM`, `YARP`, `icub-main` and some others) as they are.
